pkg/controller/ippool/pool_controller.go

```diff
@@ -0,0 +1,332 @@
+	// the Installation resource.
+	toDelete := []client.Object{}
+	for cidr, pool := range ourPools {
+		reqLogger.WithValues("cidr", cidr).Info("Checking if pool is still valid")
+		found := false
+		for _, p := range installation.Spec.CalicoNetwork.IPPools {
+			if p.CIDR == cidr {
+				found = true
+				break
+			}
+		}
+		if !found {
+			// This pool needs to be deleted. We only ever send deletes via the API server,
+			// since deletion requires rather complex logic. If the API server isn't available,
+			// we'll instead just mark the pool as disabled temporarily.
+			reqLogger.WithValues("cidr", cidr).Info("Pool needs to be deleted")
+			if apiAvailable {
+				// v3 API is available - send a delete request.
+				v3res, err := v1ToV3(&pool)
+				if err != nil {
+					r.status.SetDegraded(operator.ResourceValidationError, "error handling IP pool", err, reqLogger)
+					return reconcile.Result{}, err
+				}
+				toDelete = append(toDelete, v3res)
+			} else {
+				// v3 API is not available. Just mark the pool as disabled so that new allocations
+				// don't come from this pool. We'll delete it once the API server is available.
+				pool.Spec.Disabled = true
+				toCreate = append(toCreate, &pool)
+			}
+		}
+	}
+
+	// Update pools in the cluster.
+	// TODO: This is bypassing a lot of validaiton! For example:
+	// - Verify that the pool doesn't overlap with an existing pool.
+	// - Verify that the block size hasn't changed.
+	// - Verify that the pool doesn't overlap with existing IPAM blocks.
+	// We should add this validation to productize this. Ideally, we'd go through the API server (and maybe we should at steady-state).
+	// We only need to use the crd.projectcalico.org API for bootstrapping.
+	handler := utils.NewComponentHandler(log, r.client, r.scheme, installation)
+
+	passThru := render.NewPassthrough(toCreate...)
+	if err := handler.CreateOrUpdateOrDelete(ctx, passThru, nil); err != nil {
+		r.status.SetDegraded(operator.ResourceUpdateError, "Error creating / updating IPPools", err, log)
+		return reconcile.Result{}, err
+	}
+	delPassThru := render.NewDeletionPassthrough(toDelete...)
+	if err := handler.CreateOrUpdateOrDelete(ctx, delPassThru, nil); err != nil {
+		r.status.SetDegraded(operator.ResourceUpdateError, "Error deleting / updating IPPools", err, log)
+		return reconcile.Result{}, err
+	}
+
+	// Tell the status manager that we're ready to monitor the resources we've told it about and receive statuses.
+	r.status.ReadyToMonitor()
+
+	// We can clear the degraded state now since as far as we know everything is in order.
+	r.status.ClearDegraded()
+
+	if !r.status.IsAvailable() {
+		// Schedule a kick to check again in the near future. Hopefully by then
+		// things will be available.
+		return reconcile.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	// Update status to include the full set of IP pools. This tells the core controller
+	// that it's good to start installing Calico.
```

Comment:
The core controller could get the status of the IPPool controller to know if it is "good to start installing Calico".

Reply:
I've updated so that the core controller now waits for IPPools to actually exist in the API server.